sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.catalyst.optimizer.JoinSelectionHelper
+import org.apache.spark.sql.catalyst.plans.logical.{HintInfo, Join, JoinStrategyHint, LogicalPlan, NO_BROADCAST_HASH, SHUFFLE_HASH}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * This optimization rule includes two join selection:
+ *   1. detects a join child that can not broadcast and every partition size less than local map
+ *      threshold. Adds a shuffled_hash hint to force being shuffle hash join.
+ *   2. detects a join child that has a high ratio of empty partitions.
+ *      Adds a no_broadcast_hash_join hint to avoid it being broadcast.
+ */
+object DynamicJoinSelection extends Rule[LogicalPlan] with JoinSelectionHelper {
+
+  private def shouldDemoteBroadcastHashJoin(stage: ShuffleQueryStageExec): Boolean = {
+    val mapStats = stage.mapStats.get
+    val partitionCnt = mapStats.bytesByPartitionId.length
+    val nonZeroCnt = mapStats.bytesByPartitionId.count(_ > 0)
+    partitionCnt > 0 && nonZeroCnt > 0 &&
+      (nonZeroCnt * 1.0 / partitionCnt) < conf.nonEmptyPartitionRatioForBroadcastJoin
+  }
+
+  private def preferShuffledHashJoin(plan: LogicalPlan, stage: ShuffleQueryStageExec): Boolean = {
+    val localMapThreshold = conf.shuffleHashJoinLocalMapThreshold
+    !canBroadcastBySize(plan, conf) &&
```

Comment:
So what we really need is a hint that prefers shuffle hash join over SMJ, but still does broadcast if applicable.

I think it's more robust to create a new hint for this purpose, then we don't need to check canBroadcast here.

Reply:
yea, this is my origin thought of `PREFER_SHUFFLE_JOIN` that try to avoid touch broadcast join at this place. So follow this idea, it is not exclusive with `NO_BROADCAST_HASH` then I used a `JoinStrategyHintCollection` to keep these two hints.

Do you think which version is better ?

Reply:
OK I got your point now. For a certain join, we may want to demote the broadcast join, and we may also want to switch to shuffle hash join. These two can happen at the same time.

However, if we demote the broadcast join, many partitions are empty and I don't think the join algorithm matters anymore. We can just say:
1. If  `NO_BROADCAST_HASH`, we should use SMJ
2. If `PREFER_SHUFFLE_HASH`, we should use shuffle hash join if broadcast join is not applicable.

Reply:
or we can have 3 hints for different cases
1. if we want to demote and want to use shuffle hash join, then use `SHUFFLE_HASH` hint
2. if we don't want to demote and want to use shuffle hash join, then use `PREFER_SHUFFLE_HASH` hint
3. if we want to demote and don't want to use shuffle hash join, then use `NO_BROADCAST_HASH` hint

Reply:
the last idea +1